docs/okeps/okep-5094-layer2-transit-router.md

```diff
@@ -0,0 +1,757 @@
+
+```yaml
+k8s.ovn.org/node-transit-peer-ports-subnet: '{"ipv4":"100.88.0.12/30","ipv6":"fd97::8/127"}'
+```
+
+Since these subnet can be shared between layer2 networks they don't need to be qualified by network name, same as layer3 swith port ifaddr.
+
+### Implementation Details
+
+A [PoC PR](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5067) exist with basic implementation to make all CI pass and as input
+for this document.
+
+#### At ovnkube-node
+
+Create a helper that is able to derive the peer IPs from the node annotation
+for the transit peers node subnet.
+
+The layer2 controller should do the following:
+- Adapt and call the syncNodeClusterRouterPort from layer3 that creates the LRP that connectes to the switch at ovn_cluster_router
+- Pass ovn_cluster_router as cluster_name to the gateway init functions
+- Remove from gateway.go the code that was attaching GR to layer2 switch
+- Add to gateway.go the code that connects GR to ovn_cluster_router using the peer ports
+- Change egressip.go so routes are configured at ovn_cluster_router instead of gateway router.
+- Change the ZoneInterConnect handler to be able to add the transit router remote ports to ovn_cluster_router
+- In general re-use as much Layer3 code as possible since these makes Layer2 topology similar to it.
+
+#### At ovnkube-control-plane
+
+At cluster manager at the zuneCluster controller calculate the peers node subnet
+and annotate the node.
+
+To calculate the peer node subnet we have two options:
+- Derive it from the node-id (same we do for transit switch addrs):
+  - good:
+    - No need to maintain the allocator lifecycle
+    - It's possible to entertain the idea of not annotating the node since the source of truth is the node-id
+  - bad:
+    - Add complexity in the form of node-id to subnet derivation code
+    - Make the transit peers node subnet dependant on node-id
+- Create a new subnet allocator that will be re-sync on restarts and allocate 
+  the subnet with it:
+  - good:
+    - Is simpler since it's matter of a new subnet allocator
+    - The transit peers node subnet is no longer dependant on node-id
+  - bad:
+    - Need to maintain the allocator lifecycle
+    - Use more memory
+
+#### Gateway reconciliation
+
+After this topology is configured ovn-kubernetes need to make sure that the new node agnostic gateway are
+configured and that the old gateway is gone.
+
+For IPv4 a garp need to be generated with the gateway from ovn_cluster_router this will overwrite 
+the mac address from the .1 address at the neighbors cache.
+
+For IPv6 two maint things need to be done
+- Send a unsolicited RA to configure the ovn_cluster_router gateway
+- Send one unsolicited RA per node with lifetime 0 to remove the gateways from GRs.
+
+#### Upgrade cleanups
+
+The implementation should handle the ports, routes, policies and NAT that is no longer
+at the gateway router or layer2 switch and remove it so there is nothing dangling from previous
```

Comment:
s/at the/configured on the/

s/remove it/remove them/

s/nothing dangling from previous/no stale configuration from the previous/